From 3a3263bbc357c28e29f1b7c6ada51792ab2e5500 Mon Sep 17 00:00:00 2001
From: Lei Xu <lei.xu@nxp.com>
Date: Thu, 12 Jun 2025 09:57:04 +0800
Subject: [PATCH 29/56] LF-15696: arm64: dts: frdm-imx91: create media graph
 for camera MT9M114

Add the parallel csi and isi nodes for FRDM-IMX91 platform.
Create media graph to setup the endpoint connection between
ISI, CSI, MT9M114 and enable them by default.

Signed-off-by: Lei Xu <lei.xu@nxp.com>
Reviewed-by: Guoniu Zhou <guoniu.zhou@nxp.com>
---
 arch/arm64/boot/dts/freescale/Makefile        |   3 +-
 .../freescale/imx91-11x11-frdm-mt9m114.dts    | 201 ++++++++++++++++++
 2 files changed, 203 insertions(+), 1 deletion(-)
 create mode 100644 arch/arm64/boot/dts/freescale/imx91-11x11-frdm-mt9m114.dts

diff --git a/arch/arm64/boot/dts/freescale/Makefile b/arch/arm64/boot/dts/freescale/Makefile
index 9c823946e642..523ff1d93238 100644
--- a/arch/arm64/boot/dts/freescale/Makefile
+++ b/arch/arm64/boot/dts/freescale/Makefile
@@ -551,7 +551,8 @@ dtb-$(CONFIG_ARCH_MXC) += imx93-11x11-evk-pmic-pf0900-ld.dtb
 dtb-$(CONFIG_ARCH_MXC) += imx93-11x11-evk-pmic-pf0900-rpmsg.dtb
 dtb-$(CONFIG_ARCH_MXC) += imx93-11x11-evk-pmic-pf0900-rpmsg-lpv.dtb
 
-dtb-$(CONFIG_ARCH_MXC) += imx91-11x11-frdm.dtb
+dtb-$(CONFIG_ARCH_MXC) += imx91-11x11-frdm.dtb \
+			  imx91-11x11-frdm-mt9m114.dtb
 
 imx91-11x11-frdm-tianma-wvga-panel-dtbs := imx91-11x11-frdm.dtb imx91-11x11-frdm-tianma-wvga-panel.dtbo
 dtb-$(CONFIG_ARCH_MXC) += imx91-11x11-frdm-tianma-wvga-panel.dtb
diff --git a/arch/arm64/boot/dts/freescale/imx91-11x11-frdm-mt9m114.dts b/arch/arm64/boot/dts/freescale/imx91-11x11-frdm-mt9m114.dts
new file mode 100644
index 000000000000..66b9deb0d846
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx91-11x11-frdm-mt9m114.dts
@@ -0,0 +1,201 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2025 NXP
+ */
+
+#include <dt-bindings/media/video-interfaces.h>
+#include "imx91-11x11-frdm.dts"
+
+/* Disable lpspi3 node since pin conflict. */
+/delete-node/ &lpspi3;
+
+/ {
+	reg_vexp_3v3: regulator-vexp-3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "VEXP_3V3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		gpio = <&pcal6524 2 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+		regulator-always-on;
+	};
+
+	reg_mt9m114_dovdd: regulator-mt9m114-dovdd {
+		compatible = "regulator-fixed";
+		regulator-name = "mt9m114-dovdd";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+		startup-delay-us = <0>;
+		off-on-delay-us  = <1000>;
+		gpio = <&pca9538 0 GPIO_ACTIVE_HIGH>;
+		vin-supply = <&reg_vexp_3v3>;
+		enable-active-high;
+	};
+
+	reg_mt9m114_avdd: regulator-mt9m114-avdd {
+		compatible = "regulator-fixed";
+		regulator-name = "mt9m114-avdd";
+		regulator-min-microvolt = <2800000>;
+		regulator-max-microvolt = <2800000>;
+		startup-delay-us = <200>;
+		off-on-delay-us  = <500>;
+		gpio = <&pca9538 1 GPIO_ACTIVE_HIGH>;
+		vin-supply = <&reg_vexp_3v3>;
+		enable-active-high;
+	};
+
+	reg_mt9m114_dvdd: regulator-mt9m114-dvdd {
+		compatible = "regulator-fixed";
+		regulator-name = "mt9m114-dvdd";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+		startup-delay-us = <500>;
+		off-on-delay-us  = <200>;
+		gpio = <&pca9538 2 GPIO_ACTIVE_HIGH>;
+		vin-supply = <&reg_vexp_3v3>;
+		enable-active-high;
+	};
+
+	reg_mt9m114_extclk: regulator-mt9m114-extclk {
+		compatible = "regulator-fixed";
+		regulator-name = "mt9m114-extclk";
+		startup-delay-us = <10000>;
+		off-on-delay-us  = <200>;
+		vin-supply = <&reg_vexp_3v3>;
+		gpio = <&pca9538 3 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+
+	reg_mt9m114_ctrl_9509: regulator-mt9m114-ctrl-9509 {
+		compatible = "regulator-fixed";
+		regulator-name = "mt9m114-ctrl-9509";
+		gpio = <&pca9538 4 GPIO_ACTIVE_HIGH>;
+		vin-supply = <&reg_vexp_3v3>;
+		enable-active-high;
+		regulator-always-on;
+	};
+
+	reg_mt9m114_ctrl_4t245: regulator-mt9m114-ctrl-4t245 {
+		compatible = "regulator-fixed";
+		regulator-name = "mt9m114-ctrl-4t245";
+		startup-delay-us = <10000>;
+		off-on-delay-us  = <200>;
+		gpio = <&pca9538 5 GPIO_ACTIVE_HIGH>;
+		vin-supply = <&reg_vexp_3v3>;
+		enable-active-low;
+	};
+
+	reg_mt9m114_ctrl_16t245: regulator-mt9m114-ctrl-16t245 {
+		compatible = "regulator-fixed";
+		regulator-name = "mt9m114-ctrl-16t245";
+		gpio = <&pca9538 6 GPIO_ACTIVE_HIGH>;
+		vin-supply = <&reg_vexp_3v3>;
+		enable-active-low;
+		regulator-always-on;
+	};
+};
+
+&iomuxc {
+	pinctrl_parallel_csi: ci_pi {
+		fsl,pins = <
+			MX91_PAD_GPIO_IO01__MEDIAMIX_CAM_DATA0		0xb9e
+			MX91_PAD_GPIO_IO07__MEDIAMIX_CAM_DATA1		0xb9e
+			MX91_PAD_GPIO_IO08__MEDIAMIX_CAM_DATA2		0xb9e
+			MX91_PAD_GPIO_IO09__MEDIAMIX_CAM_DATA3		0xb9e
+			MX91_PAD_GPIO_IO10__MEDIAMIX_CAM_DATA4		0xb9e
+			MX91_PAD_GPIO_IO11__MEDIAMIX_CAM_DATA5		0xb9e
+			MX91_PAD_GPIO_IO14__MEDIAMIX_CAM_DATA6		0xb9e
+			MX91_PAD_GPIO_IO15__MEDIAMIX_CAM_DATA7		0xb9e
+			MX91_PAD_GPIO_IO17__MEDIAMIX_CAM_DATA8		0xb9e
+			MX91_PAD_GPIO_IO18__MEDIAMIX_CAM_DATA9		0xb9e
+
+			MX91_PAD_GPIO_IO00__MEDIAMIX_CAM_CLK		0xb9e
+			MX91_PAD_GPIO_IO02__MEDIAMIX_CAM_VSYNC		0xb9e
+			MX91_PAD_GPIO_IO03__MEDIAMIX_CAM_HSYNC		0xb9e
+			MX91_PAD_GPIO_IO05__GPIO2_IO5			0x31e
+			MX91_PAD_GPIO_IO04__GPIO2_IO4			0x31e
+		>;
+	};
+
+	pinctrl_lpi2c8: lpi2c8grp {
+		fsl,pins = <
+			MX91_PAD_GPIO_IO12__LPI2C8_SDA		0x40000b9e
+			MX91_PAD_GPIO_IO13__LPI2C8_SCL		0x40000b9e
+		>;
+	};
+};
+
+&isi {
+	status = "okay";
+
+	port {
+		isi_in: endpoint {
+			remote-endpoint = <&parallel_csi_out>;
+		};
+	};
+};
+
+&parallel_csi {
+	status = "okay";
+
+	ports {
+		port@0 {
+			reg = <0>;
+
+			parallel_csi_in: endpoint {
+				remote-endpoint = <&mt9m114_ep>;
+			};
+		};
+
+		port@1 {
+			reg = <1>;
+
+			parallel_csi_out: endpoint {
+				remote-endpoint = <&isi_in>;
+			};
+		};
+	};
+};
+
+&lpi2c8 {
+	clock-frequency = <400000>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_lpi2c8>;
+	status = "okay";
+
+	pca9538: gpio@70 {
+		compatible = "nxp,pca9538";
+		reg = <0x70>;
+		gpio-controller;
+		#gpio-cells = <2>;
+		vin-supply = <&reg_vexp_3v3>;
+		status = "okay";
+	};
+
+	mt9m114: mt9m114@48 {
+		compatible = "on,mt9m114";
+		reg = <0x48>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&pinctrl_parallel_csi>;
+		reset-gpios = <&gpio2 4 GPIO_ACTIVE_HIGH>;
+		powerdown-gpios = <&gpio2 5 GPIO_ACTIVE_HIGH>;
+		DOVDD-supply = <&reg_mt9m114_dovdd>;
+		AVDD-supply = <&reg_mt9m114_avdd>;
+		DVDD-supply = <&reg_mt9m114_dvdd>;
+		EXTCLK-supply = <&reg_mt9m114_extclk>;
+		CTRL_4T245-supply = <&reg_mt9m114_ctrl_4t245>;
+		mclk = <27000000>;
+		status = "okay";
+
+		port {
+			mt9m114_ep: endpoint {
+				remote-endpoint = <&parallel_csi_in>;
+				bus-type = <MEDIA_BUS_TYPE_PARALLEL>;
+				bus-width = <8>;
+				vsync-active = <1>;
+				hsync-active = <1>;
+				pclk-sample = <1>;
+			};
+		};
+	};
+};
-- 
2.34.1

